### PR TITLE
ci: cache playwright browser

### LINKED ISSUE #120: ci: cache playwright browser
Every action run should not install the playwright browser. That takes way too long. Look into how we can cache the browser to make the CI job run faster.

### PR DESCRIPTION
**Pull request by shawnyu5**
Closes #120

### CODE REVIEW

**Comment by shawnyu5** [comment]
It is possible to cache the playwright browser it self, however the browser requires system dependencies, that can not be cached easily. Refer to this run here: https://github.com/shawnyu5/starchart/actions/runs/4093992063

The alternative is to install the browser it self through `npx`, and system dependencies through the package manager. Which still takes too long. Refer to this run here: https://github.com/shawnyu5/starchart/actions/runs/4093992063

Going to experiment with using a docker container with playwright installed to run the tests

**Comment by shawnyu5** [comment]
Hmm, using [this](https://hub.docker.com/_/microsoft-playwright?tab=description) docker container, that contains playwright to run our CI means we would need to install docker inside the container every time. Which also involves installing stuff, and is no better than installing playwright it self...

**Comment by shawnyu5** [comment]
Since the CI run can take a little bit of time, a `concurrency` group is added to cancel previous workflow runs if a new workflow is triggered.

For example, if 2 commits were pushed in close succession. Rather than kicking off 2 separate workflow runs, the first workflow run would be cancelled in favor of the second run. Refer to these 2 workflow runs:

First commit triggering a workflow run, cancelled by the next commit: https://github.com/Seneca-CDOT/star chart/actions/runs/4094766475

Second commit, pushed right after the previous commit. Triggering another workflow run, cancelling the first: https://github.com/Seneca-CDOT/star chart/actions/runs/4094767877

**Comment by shawnyu5** [comment]
The CI run will attempt to restore playwright browser from cache. If it is successful, it will install system dependencies. If its unsuccessful, it will install playwright browser it self along with its dependencies

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -58,7 +60,6 @@ jobs:
           prettier_options: --check .
 
   Unit-Tests:
-    permissions: write-all
```
Comment (Eakam1007) [inline .github/workflows/ci.yaml:61]: Just curious, what did this code do before? Is this not needed?

**Review by Eakam1007 — DISMISSED** [review]
Looks good! It is really nice to see faster tests

Just wondering; maybe this could be addressed in another issue; is it possible to cache the dependencies as well?

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (shawnyu5) [reply .github/workflows/ci.yaml:61]: This was needed with the weird hack of canceling parallel jobs. Forgot to clean it up in the other PR.

**Comment by shawnyu5** [comment]
The dependencies are system dependencies from what I understand, it's not a single binary, like the browser, or a single directory like `node_modules`. So I dont think its fully possible to cache those.

**Comment by shawnyu5** [comment]
If you look at the logs [here](https://github.com/Seneca-CDOT/starchart/actions/runs/4100467684/jobs/7071323094#step:9:1), the dependencies are being installed through the package manager... I don't know if it's possible to cache those...

**Comment by shawnyu5** [comment]
Cleaned up some commit history

**Review by Eakam1007 — APPROVED** [review]

**Review by humphd — APPROVED** [review]
Very nice!

**Pull request merged**